drivers/adc_ng/adc_ng.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2020 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_adc_ng
+ * @{
+ *
+ * @file
+ * @brief       Implementation of common functions of the ADC API
+ *
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ * @}
+ */
+
+#include <errno.h>
+
+#include "adc_ng.h"
+
+#if (ADC_NG_NUMOF == 1) && defined(MODULE_PERIPH_ADC_NG)
```

Comment:
This prevents that a board could declare to use `adc_ng` with only one external ADC if the MCU provides that feature. This might be solved if not the MCU defines that feature but the board as for other peripherals.

Reply:
My logic here is: If the module for the peripheral ADC (with ADC NG API support) is used and in total one ADC NG backend is used, that the single backend has to be the peripheral ADC.

If I would want to use the external ADC only and disable the module for the internal ADC, `ADC_NG_NUMOFF == 1` would be true, but `MODULE_PERIPH_ADC_NG` should not be defined.

Reply:
(I just hit resolve here whenever I think that the issue is resolved to get a better overview of the still open todos. This should not mean that I want to stop the discussion; so just hit unresolve if I was to fast with resolving the issue.)

Reply:
> My logic here is: If the module for the peripheral ADC (with ADC NG API support) is used and in total one ADC NG backend is used, that the single backend has to be the peripheral ADC.

The logic behind this is clear and if all MCUs implement `adc-ng` it could be OK. But if a board wants to use `adc_ng` with an external ADC, it would not be possible with this logic if the MCU does not support `adc_ng`.

Another use case could be that a board wants to use only external ADCs but no peripheral MCUs.

Another use case that is affected by this logic is the test application for an external ADC. This test application must define an external ADC independently of the board and the MCU's support of `adc_ng`.

> If I would want to use the external ADC only and disable the module for the internal ADC, `ADC_NG_NUMOFF == 1` would be true, but `MODULE_PERIPH_ADC_NG` should not be defined.

It would be fine, if it would be possible to disable `periph_adc_ng`. But it seems that it is not possible to disable a module that is pulled in by `FEATURES_PROVIDED`
```
DISABLE_MODULE=periph_adc_ng make BOARD=arduino-mega2560 -C tests/driver_ads101x_adc_ng
```
gives
```
Required modules were disabled using DISABLE_MODULE: periph_adc_ng
```
Therefore, I unresolved this for the moment.

Reply:
> This might be solved if not the MCU defines that feature but the board as for other peripherals.

As I understand it, the concept of board definitions in RIOT is that the decision whether or not to use a peripheral is made by the board and not by the MCU, except for internal MCU functions such as the RTC.

That is, even if the MCU supports a peripheral, the board uses `FEATURES_PROVIDED` to enable it. The reason is, that a certain peripherals might be not broken on the board.

For example, board definition in `boards/common/arduino-atmega` defines the peripherals used by these boards, including `periph_adc`.

Reply:
`FEATURES_PROVIDED` does not enable a feature. It allows enabling features. If features are actually used depends actually on the application; only as dependency to application requested modules or features other modules or features get pulled in.

Note that the above logic checks for use of the module `periph_adc_ng`. On boards that do not support peripheral ADC(s) with ADC NG backend, the module `periph_adc_ng` would not be inplemented and therefore would never be used. (Unless the application explicitly adds that module to `USE_MODULE`, which in turn would result in a build failure.)

Reply:
> `FEATURES_PROVIDED` does not enable a feature. It allows enabling features. If features are actually used depends actually on the application; only as dependency to application requested modules or features other modules or features get pulled in.
>
> Note that the above logic checks for use of the module `periph_adc_ng`

Yeah, that's clear. But as you explained, in the case of the `periph_*` features, it implicitly uses the corresponding `periph_*` module.

In fact, the logic prevents me to compile my `ads101x` test application for (https://github.com/RIOT-OS/RIOT/pull/13247#issuecomment-586141439) for ATmega boards. 

Since the `ads101x` test application has to work indpendent on whether the MCU supports `adc_ng` or not, it simulates a board definition with only one `ads101x`. Since `cpu/atmega_common` provides the `periph_adc_ng` feature, module `periph_adc_ng` is pulled in and the logic gives true. Unfortunately, `periph_adc_ng` can be neither blacklisted nor disabled.

A workaround might be that the `ads101x` test application defines two `ads101x` devices. But if the user has only one, application would crash during initialization of the second `ads101x`. A hack might be to define the same `ads101x` twice. A double initialization wouldn't be a problem.

Reply:
I changed `drivers/Makefile.dep` to no longer add `FEATURES_OPTIONAL := periph_adc_ng`. For that `adc_ng_default` is now used. This means that the `periph_adc_ng` feature should not be automatically used even if provided, unless the pseudo module `periph_adc_default` is used.

The test should now compile with the `ads101x` for ATmega boards if you replace `USEMODULE += adc_ng_default` by `USEMODULE += adc_ng ads101x`.

Reply:
:+1: This change did the trick.